assets/js/study.js

```diff
@@ -173,27 +175,79 @@ function give_hints(access, once) {
     ground.setShapes(shapes);
 }
 
+function capitalize_first_letter(word) {
+    return word.charAt(0).toUpperCase() + word.slice(1);
+}
+
+function create_comment(container_div, response_num, move, decorate_color) {
+    var comment_div = document.createElement('div');
+    comment_div.id = 'comment' + response_num;
+
+    var bold_text_elem = document.createElement('div');
+    bold_text_elem.id = comment_div.id + "_bold";
+    bold_text_elem.className = "bold";
+
+    var comment_text_elem = document.createElement('div');
+    comment_text_elem.id = comment_div.id + "_text";
+    comment_text_elem.className = "text";
+
+    comment_div.appendChild(bold_text_elem);
+    comment_div.appendChild(comment_text_elem);
+    container_div.appendChild(comment_div);
+
+    let response_color = capitalize_first_letter(turn_color(chess));
+    let move_color = capitalize_first_letter(non_turn_color(chess));
+    let ext_san = tree_get_node_string(move);
+
+    let text = unescape_string(move.comments[0].text.trim());
+    let bold_text = response_num == undefined ?
+        (decorate_color ? move_color : "") + " " + ext_san + ":"
+        : response_color + " " + i18n.response + " " + ext_san + ":" ;
+
+    set_text(comment_div.id, text, { bold_text: bold_text });
+}
+
+function create_comment_list(container_id, current_move, response_moves) {
+    let container_div = document.getElementById(container_id);
+    container_div.innerHTML = '';
+
+    if (current_move != undefined) {
+        // Don't decorate the current move with the color if it's the only comment
+        let decorate_color = response_moves.length > 0;
+        create_comment(container_div, undefined, current_move, decorate_color);
+    }
+
+    for (let [i, move] of response_moves.entries()) {
+        create_comment(container_div, i, move, true);
+    }
+}
+
 /*
- * Display comments for moves that need hints and the current move
+ * Display comments, either only for the current move, or for the opposite side's responses as well
+ * if that option is turned on.
  */
 function display_comments(access) {
     let cm = tree_get_node(access);
-    let comment = "";
-    let to_hint = tree_children_filter_sort(access, {filter: need_hint});
-    for (let m of to_hint) {
-        if (m.comments != undefined && m.comments[0] != undefined && m.comments[0].text != undefined) {
-            comment += m.comments[0].text.trim() + "\n";
-        }
-    }
+    let current_move = undefined;
+    let response_moves = [];
 
-    // give comment on the current move if there are move to give hint to
-    if (to_hint.length > 0 && cm.comments != undefined && cm.comments[0] != undefined && cm.comments[0].text != undefined) {
-        comment += cm.comments[0].text.trim();
+    // Get current move if it has a comment
```

Comment:
The comments should only be displayed when a arrow hint is to be displayed. Otherwise its just a spoiler.

E.g. this move was played enough times so no hint is given, but the comment still spoils the right move.
![image](https://user-images.githubusercontent.com/26939450/204079601-1c475a23-06b5-4e3e-a4d9-3f4128f88f68.png)

Reply:
Well..  I had actually intentionally changed that. But yeah I see what you mean. I think my thinking was that even if a comment is visible, it only provides some background on the position, maybe describes the name of the position, or has a longer description like in your screenshot above. It would spoil the move if you read the comment, you're right. But my thinking is that if you read the text is has to be a conscious decision, so it should be easy enough to avoid it if you don't want the help. An arrow that shows up when you don't want to see it is a lot harder to "unsee", so that's why I personally think it's much more important to make sure the user doesn't see the arrow hints than the comment hints. But, you're right, it might give a hint to the user and spoil the move if the comment is easy enough to read and interpret when you don't want to. 

 I'll change it back if you want to, no problem. Let me know what you decide!

Reply:
I would prefer to revert it back.

But the idea gave me another thought: maybe to create a new review mode.

1. Unconditionally: Show all arrows
2. Unconditionally: Show all comments
3. Show the PGN on the left
3.1. Allow the moves in the pgn to be clickable like on lichess to go to the move.
4. With the addition of statistics collection
4.1. Highlight "weak" moves, where the most errors were made in training mode in the pgn
4.2. When on the "weak" move on the board somehow make it obvious that more attention should be given to this move

Of course out of scope for this pr but just commenting as a idea for the future.

Reply:
Hmm.. It gets complicated when you think about it. The comments for the played move is usually the opponents moves, unless you have a slow move delay I suppose, in case I think your own played response's comment would be "the played move" (haven't tested that to be honest, and I'm out hiking today :)). But also, would people really read their own response's comment in the second or so after the response was played? I don't think so. 

So, if we connect the displaying of arrows to the displaying of comments, then you won't see the opponents (the played move's) comments, because there are never any arrows shown for the opponent's moves. I guess that's how it was before though. But it's a little strange I think. Turning off your own arrows turns of the opponents comments. 

How about:
Comments: played move (always).           (Ie the opponents move in practicality)
Comments: all moves (always)
Comments: played move always + responses with arrows
Comments: responses with arrows

The problem I have is finding a short enough text string to describe these modes.

Or, another option altogether:

We **always** display comments for the opponent's moves (not only when the played move is his). And we allow the option to control if **your** move's comments are shown or not:

Comments: all your moves
Comments: for arrowed moves

But maybe just with better wording.

Or leave it as is. :)

Reply:
Wait a minute.... there's another aspect to this altogether. Maybe, just maybe, apart from comments on the move the opponent just made, maybe the user doesn't want to see comments for possible upcoming responses, but comments on the **one single** move the user played one move back in time. I mean it's only **that** comment that describes the position we ended up with. The problem was in most cases the opponent (ai) then played his move instantly and without delay, so we never had a chance to read the comment of the position we gave ourselves. 

Correct me if I am wrong, but I don't think that is how the code worked before. So making such a change would break backwards compatibility.

Man it gets complicated. :)

Reply:
> maybe the user doesn't want to see comments for possible upcoming responses

I don't get this. I think it would be weird to be one move "behind". 

From https://lichess.org/study/Jn2JyhRT/6nJ8VQoA :

From whites perspective currently it is what I would expect (and what is implemented) while arrows are shown:
![image](https://user-images.githubusercontent.com/26939450/204099254-3d406302-d8fc-47d5-b579-1b74b764b2bc.png)

I think this should be the default while arrows are shown, with no arrows in the position only the opponent moves should be shown.
And the comment option should then be on if comments are shown in this way, or no comments are shown at all.

Reply:
Yeah after thinking more about it, I think you're definitely right!

I think the key here is whether most comments talks mostly about the current position or any follow-up moves the other side may have. And it's probably more likely that a comment focuses on the current position than future responses, so for that reason it's better to show current and future comments.

So I like your ideas and will update the PR accordingly.

But... to achieve what you want of not spoiling any information, I think we need to hide **all** comments, including the "played move". Consider this simple example. If for example the comment on 1.e4 says "Black plays the most common opening, and white can respond with e5 or c5" then that definitely spoils the position for black.

Maybe the whole idea of having an option for controlling whether comments from one or both sides are displayed is a bit of overkill. Especially now that it's clearer visually what move a comment belongs to.

So how about this: I fix the other issues you mentioned below. And we also keep the slightly more refined visual changes in how comments are displayed. And then make the option as follows instead:

Comments: when arrows show
Comments: always on
Comments: always off

Personally I will practice with arrows always off, and turn them on using the Hints link, but I always want the comments showing. And the suggestion above allows me to get that.

Reply:
I uploaded a new commit to modify these things.

So: 

- Removed the option to choose between "Comments: played move" and "Comments: played move + responses".
- Replaced with option to choose between "Comments: when arrows show", "Comments: always on", and "Comments: hidden".
- Not showing comments entered before the first move in the PGN. I think that should be a reasonable solution.
- Not calling the opening moves "responses".